StreamChatClient_v3/APIClient/RequestEncoder.swift

```diff
@@ -0,0 +1,192 @@
+//
+// RequestEncoder.swift
+// Copyright © 2020 Stream.io Inc. All rights reserved.
+//
+
+import Foundation
+
+/// On object responsible for creating a `URLRequest`, and encoding all required and `Endpoint` specific data to it.
+protocol RequestEncoder {
+    /// A delegate the encoder uses for obtaining the current `connectionId`.
+    ///
+    /// Trying to encode an `Endpoint` with the `requiresConnectionId` set to `true` without setting the delegate
+    var connectionIdProviderDelegate: ConnectionIdProviderDelegate? { get set }
+    
+    /// Asynchronously creates a new `URLRequest` with the data from the `Endpoint`. It also adds all required data
+    /// like an api key, etc.
+    ///
+    /// - Parameters:
+    ///   - endpoint: The `Endpoint` to be encoded.
+    ///   - completion: Called when the encoded `URLRequest` is ready. Called with en `Error` if the encoding fails.
+    func encodeRequest<ResponsePayload: Decodable>(for endpoint: Endpoint<ResponsePayload>,
+                                                   completion: @escaping (Result<URLRequest, Error>) -> Void)
+    
+    /// Creates a new `RequestEncoder`.
+    ///
+    /// - Parameters:
+    ///   - baseURL: The base URL for all requests.
+    ///   - apiKey: The app specific API key.
+    init(baseURL: URL, apiKey: APIKey)
+}
+
+/// The default implementation of `RequestEncoder`.
+struct DefaultRequestEncoder: RequestEncoder {
+    let baseURL: URL
+    let apiKey: APIKey
+    
+    weak var connectionIdProviderDelegate: ConnectionIdProviderDelegate?
+    
+    func encodeRequest<ResponsePayload: Decodable>(for endpoint: Endpoint<ResponsePayload>,
+                                                   completion: @escaping (Result<URLRequest, Error>) -> Void) {
+        var request: URLRequest
+        
+        do {
+            // Prepare the URL
+            var url = try encodeRequestURL(for: endpoint)
+            url = try url.appendingQueryItems(["api_key": apiKey.apiKeyString])
+            
+            // Create a request
+            request = URLRequest(url: url)
+            request.httpMethod = endpoint.method.rawValue
+            
+            // Encode endpoint-specific query items
+            if let queryItems = endpoint.queryItems {
+                try encodeJSONToQueryItems(request: &request, data: queryItems)
+            }
+            
+            try encodeRequestBody(request: &request, endpoint: endpoint)
+            
+        } catch {
+            completion(.failure(error))
+            return
+        }
+        
+        if endpoint.requiresConnectionId {
+            assert(connectionIdProviderDelegate != nil,
```

Comment:
We have `log.assert` for this

Reply:
fixed in 273c18f407b480b1bc3c9932299bcfad0415d286